File: Programming Study/DataStructure with JAVA/DataStructure/dataStructure/src/Practice/QuickSort.java
package Practice;

public class QuickSort {
     // 퀵 정렬
     public static void quickSort(int [] a, int p, int r){
        // p: 정렬할 배열의 시작 위치, r: 정렬할 배열의 마지막 위치 = 피벗
        if(p<r){
            //기준원소를 기준으로 작은 원소, 큰 원소를 각각 나누어 분할
            int pivot = partition(a, p, r);
            //왼쪽 부분배열(피벗보다 작은 값들로 구성된)을 정렬
            quickSort(a, p, pivot-1);
            //오른쪽 부분배열(피벗보다 큰 값들로 구성된)을 정렬
            quickSort(a, pivot+1, r);
        }
    }

    // 분할 알고리즘
    public static int partition(int[] a, int p, int r){
        int x = a[r]; // 기준 원소는 배열의 마지막 원소
        int i = p-1;
        for(int j=p; j<=r-1; j++){
            if(a[j] < x){
                swap(a, ++i, j); 
            }
        }
        // a[r] <-> a[i+1]
        swap(a, r, i+1);
        //partition(a, p, r) 호출 후 배열의 상태를 출력
        System.out.printf("a[%d] : %d => ", r, x);
        print(a);

        // 피벗의 위치 반환
        return i+1;
    }
    public static void main(String[] args) {
        
    }

    //기본 메소드
    public static void swap(int[] a, int i, int j){
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    public static void print(int[] a){
        for(int item:a){
            System.out.print(item + " ");
        }
        System.out.println();
    }
}

File: Data Structure and Algorithm/Practice/FinalTerm/QuickSort.py
def quick_sort(A, left, right):
    if left < right:
        q = partition(A, left, right)
        quick_sort(A, left, q-1)
        quick_sort(A, q+1, right)

def partition(A, left, right):
    pivot = A[left]
    low = left + 1
    high = right
    
    while (low < high): # low와 high가 역전되지 않는 한 반복
        while low <= right and A[low] <= pivot:
            low += 1 # A[low] <= pivot 이면 low를 오른쪽으로 진행
        
        while high >=left and A[high] > pivot:
            high -= 1 # A[high] > pivot 이면 high를 왼쪽으로 진행

        if low < high:
            A[low], A[high] = A[high], A[low]

    A[left], A[high] = A[high], A[left]
    return high

    data = [71, 49, 92, 55, 38, 28, 72, 53]
    quick_sort(data)





